File: blog/排序算法整理/selection-sort/selectionSort.go
package main

import "fmt"

func main() {
	data := []int{1, 2, 90, 29, 32, 17, 51, 100, 99, 3, 9, 8, 6, 2}
	selectionSort(data, "ASC")
	fmt.Println("ASC", data)
	selectionSort(data, "DESC")
	fmt.Println("DESC", data)
}

func selectionSort(nums []int, orderType string) {
	n := len(nums)
	for i := range nums[1:] {
		tempIdx := i
		if orderType == "ASC" {
			for j := i + 1; j < n; j++ {
				if nums[j] < nums[tempIdx] {
					tempIdx = j
				}
			}
		} else if orderType == "DESC" {
			for j := i + 1; j < n; j++ {
				if nums[j] > nums[tempIdx] {
					tempIdx = j
				}
			}
		}
		nums[i], nums[tempIdx] = nums[tempIdx], nums[i]
	}
}

File: blog/排序算法整理/selection-sort/selectionSort.js
function selectionSort (arr, orderType) {
  let len = arr.length
  for (let i = 0; i < len - 1; i++) {
    let tempIdx = i
    let temp
    if (orderType === 'ASC') {
      // 寻找最小的数
      for (let j = i + 1; j < len; j++) {
        if (arr[j] < arr[tempIdx]) {
          tempIdx = j
        }
      }
    } else if (orderType === 'DESC') {
      // 寻找最大的数
      for (let j = i + 1; j < len; j++) {
        if (arr[j] > arr[tempIdx]) {
          tempIdx = j
        }
      }
    }
    // 交换元素
    temp = arr[i]
    arr[i] = arr[tempIdx]
    arr[tempIdx] = temp
  }
  return arr
}

const arr = [1, 2, 90, 29, 32, 17, 51, 100, 99, 3, 9, 8, 6, 2]
console.log(selectionSort(arr, 'ASC'))
console.log(selectionSort(arr, 'DESC'))
